src/include/sof/audio/format_generic.h

```diff
@@ -10,44 +10,37 @@
 
 /* Saturation inline functions */
 
+static inline int32_t sat_generic(int64_t x, int64_t max_val, int64_t min_val);
+
 static inline int32_t sat_int32(int64_t x)
 {
-	if (x > INT32_MAX)
-		return INT32_MAX;
-	else if (x < INT32_MIN)
-		return INT32_MIN;
-	else
-		return (int32_t)x;
+	return sat_generic(x, INT32_MAX, INT32_MIN);
 }
 
 static inline int32_t sat_int24(int32_t x)
 {
-	if (x > INT24_MAXVALUE)
-		return INT24_MAXVALUE;
-	else if (x < INT24_MINVALUE)
-		return INT24_MINVALUE;
-	else
-		return x;
+	return sat_generic(x, INT24_MAXVALUE, INT24_MINVALUE);
 }
 
 static inline int16_t sat_int16(int32_t x)
 {
-	if (x > INT16_MAX)
-		return INT16_MAX;
-	else if (x < INT16_MIN)
-		return INT16_MIN;
-	else
-		return (int16_t)x;
+	return (int16_t)sat_generic(x, INT16_MAX, INT16_MIN);
 }
 
 static inline int8_t sat_int8(int32_t x)
 {
-	if (x > INT8_MAX)
-		return INT8_MAX;
-	else if (x < INT8_MIN)
-		return INT8_MIN;
-	else
-		return (int8_t)x;
+	return (int8_t)sat_generic(x, INT8_MAX, INT8_MIN);
+}
+
+static inline int32_t sat_generic(int64_t x, int64_t max_val, int64_t min_val)
+{
+	int64_t mask_overflow = (max_val - x) >> 63;
+	int64_t mask_underflow = (x - min_val) >> 63;
+
+	/* Apply masks to selectively replace x with min or max values, or keep x as is. */
+	x = (x & ~mask_overflow) | (max_val & mask_overflow);
+	x = (x & ~mask_underflow) | (min_val & mask_underflow);
```

Comment:
hm, I'm counting 12 64-bit operations here. Is this really faster than 2 `if`s?

Reply:
@ShriramShastry sorry, I think there's a mistake in your analysis, at least in the first screenshot. In it you show the execution of the `sat_int32()` function which of course is much smaller now. But we need to look at the `sat_generic()` function which I'm pretty sure will be larger and possibly slower than the present version.

Reply:
I think running a pipeline on real HW before and after will show the change in a E2E use case and show us an overall change

Reply:
Measurement done on `hifi5_ss_spfpu_7`

_PR GitHub code - Generic Saturation `Bitwise -32-bit` function_

![image](https://github.com/thesofproject/sof/assets/61411947/50a65f41-9eb8-49f0-84d4-7614f28fb208)

_Original GitHub code - Generic Saturation `Branch -32` function_
![image](https://github.com/thesofproject/sof/assets/61411947/24841529-a5d4-4a72-a029-69b4e877fd8e)

_Xtensa setting_
![image](https://github.com/thesofproject/sof/assets/61411947/36d8960f-b584-418d-af62-edaa1b1889ba)


_Profiler's Output_
![image](https://github.com/thesofproject/sof/assets/61411947/cd3ca6c2-3e9c-4b31-b04a-f459f0a636b9)

_RSR.CCOUNT  -  Read Special Register_ 
![image](https://github.com/thesofproject/sof/assets/61411947/5c94ce4b-b35e-4bb1-89e5-dcfce382fcb4)

Reply:
@ShriramShastry sorry, I dont need the full simulator logs but, but just a abbreviated clip that shows before and after (as I cant tell what numbers I should be looking at above).

Reply:
OK,

In the image below, I captured performance data in terms of cycle count, and the profiler will return similar results.

Previously, the Original Function in GitHub used the branching `if-else` approach to calculate saturation. After that, the modified function under review calculates saturation using `bitwise` arithmetic.

_**(1) Generic_HiFi5**_

![image](https://github.com/thesofproject/sof/assets/61411947/2d6dbbc2-d5fa-4d11-94bf-a45c3738fa7f)


_**(2) Generic_HiFi4**_

![image](https://github.com/thesofproject/sof/assets/61411947/6a3a06b6-192b-464a-8d5f-3503a87b28b2)

_**(3) Generic_HiFi3**_

![image](https://github.com/thesofproject/sof/assets/61411947/a896d2a9-b358-4a4f-8ad2-5fac9837bbfc)

_**(4) ACE_3**_

![image](https://github.com/thesofproject/sof/assets/61411947/5e786e7a-f96c-46e1-9175-38bf94423400)

_**(5) ACE_1**_

![image](https://github.com/thesofproject/sof/assets/61411947/5ba72f7a-eb00-442c-95fe-89e3b34d86b8)

**Observation** : 

The saturation function limits the input value to a certain range, specific to the bit width of the output (32-bit, 24-bit, 16-bit, 8-bit). The `Cycles (Before)` and `Cycles (After)` columns indicate the number of cycles it took to execute the function before and after some modification, respectively.

Here's a summary of the cycles before and after the modification:

For `sat_int32`, the number of cycles remained the `same` before and after the modification, with 20 cycles for each operation.
For `sat_int24`, the number of cycles `increased` after the modification, from 16 cycles before to 21 cycles after.
For `sat_int16`, the number of cycles also `increased` after the modification, from 13 cycles before to 21 cycles after.
For `sat_int8`, the number of cycles `increased` after the modification as well, from 13 cycles before to 20 cycles after.

In summary, the modification resulted in an `increase` in the number of cycles for `sat_int24`, `sat_int16`, and `sat_int8 `functions, while the `sat_int32` function's performance remained `unchanged` in terms of cycles.

Reply:
> In summary, the modification resulted in an `increase` in the number of cycles for `sat_int24`, `sat_int16`, and `sat_int8 `functions, while the `sat_int32` function's performance remained `unchanged` in terms of cycles.

Thank you, this matches my understanding, that those 64-bit calculations cannot result in better performance than a single comparison. Maybe this PR should be closed as a failed optimisation attempt